rust/rpc/src/error.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright 2017 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io;
+use std::fmt;
+
+use serde_json::{Value, Error as JsonError};
+use serde::de::{Deserializer, Deserialize};
+use serde::ser::{Serializer, Serialize};
+
+/// Errors that can occur when sending an RPC.
+#[derive(Debug)]
+pub enum Error {
+    /// An IO error occurred on the underlying communication channel.
+    IoError(io::Error),
+    /// The peer returned an error.
+    RemoteError(RemoteError),
+    /// The peer closed its connection.
+    PeerDisconnect,
+    /// The peer sent a response containing the id, but was malformed.
+    InvalidResponse,
+}
+
+/// Errors that can occur when attemping to read a message.
+#[derive(Debug)]
+pub enum ReadError {
+    /// An error occured in the underlying stream
+    Io(io::Error),
+    /// The message was not valid JSON.
+    Json(JsonError),
+    /// The message was not a JSON object.
+    NotObject,
+    /// The peer closed the connection.
+    Disconnect,
+}
+
+impl fmt::Display for ReadError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ReadError::Io(ref err) => write!(f, "I/O Error: {:?}", err),
+            ReadError::Json(ref err) => write!(f, "JSON Error: {:?}", err),
+            ReadError::NotObject => write!(f, "JSON message was not an object."),
+            ReadError::Disconnect => write!(f, "Peer closd the connection."),
+        }
+    }
+}
+
+impl From<JsonError> for ReadError {
+    fn from(err: JsonError) -> ReadError {
+        ReadError::Json(err)
+    }
+}
+
+impl From<io::Error> for ReadError {
+    fn from(err: io::Error) -> ReadError {
+        ReadError::Io(err)
+    }
+}
+
+//TODO: code review discussion: do we want, in general, to support these
```

Comment:
Result of discussion: for request, we should return an error and then close the connection (without disrupting core). For notification, we should log an error and close the connection.